### PR TITLE
[SPARK-17952][SQL] Nested Java beans support in createDataFrame

### PR DESCRIPTION
**Pull request by michalsenkyr**
## What changes were proposed in this pull request?

When constructing a DataFrame from a Java bean, using nested beans throws an error despite [documentation](http://spark.apache.org/docs/latest/sql-programming-guide.html#inferring-the-schema-using-reflection) stating otherwise. This PR aims to add that support.

This PR does not yet add nested beans support in array or List fields. This can be added later or in another PR.

## How was this patch tested?

Nested bean was added to the appropriate unit test.

Also manually tested in Spark shell on code emulating the referenced JIRA:

```
scala> import scala.beans.BeanProperty
import scala.beans.BeanProperty

scala> class SubCategory(@BeanProperty var id: String, @BeanProperty var name: String) extends Serializable
defined class SubCategory

scala> class Category(@BeanProperty var id: String, @BeanProperty var subCategory: SubCategory) extends Serializable
defined class Category

scala> import scala.collection.JavaConverters._
import scala.collection.JavaConverters._

scala> spark.createDataFrame(Seq(new Category("s-111", new SubCategory("sc-111", "Sub-1"))).asJava, classOf[Category])
java.lang.IllegalArgumentException: The value (SubCategory@65130cf2) of the type (SubCategory) cannot be converted to struct<id:string,name:string>
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:262)
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:238)
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:396)
  at org.apache.spark.sql.SQLContext$$anonfun$beansToRows$1$$anonfun$apply$1.apply(SQLContext.scala:1108)
  at org.apache.spark.sql.SQLContext$$anonfun$beansToRows$1$$anonfun$apply$1.apply(SQLContext.scala:1108)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
  at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
  at org.apache.spark.sql.SQLContext$$anonfun$beansToRows$1.apply(SQLContext.scala:1108)
  at org.apache.spark.sql.SQLContext$$anonfun$beansToRows$1.apply(SQLContext.scala:1106)
  at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
  at scala.collection.Iterator$class.toStream(Iterator.scala:1320)
  at scala.collection.AbstractIterator.toStream(Iterator.scala:1334)
  at scala.collection.TraversableOnce$class.toSeq(TraversableOnce.scala:298)
  at scala.collection.AbstractIterator.toSeq(Iterator.scala:1334)
  at org.apache.spark.sql.SparkSession.createDataFrame(SparkSession.scala:423)
  ... 51 elided
```

New behavior:

```
scala> spark.createDataFrame(Seq(new Category("s-111", new SubCategory("sc-111", "Sub-1"))).asJava, classOf[Category])
res0: org.apache.spark.sql.DataFrame = [id: string, subCategory: struct<id: string, name: string>]

scala> res0.show()
+-----+---------------+
|   id|    subCategory|
+-----+---------------+
|s-111|[sc-111, Sub-1]|
+-----+---------------+
```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -1100,13 +1101,24 @@ object SQLContext {
...
+    }
+    def invoke(element: Any)(tuple: (Method, DataType)): Any = tuple match {
+      case (e, structType: StructType) =>
+        val value = e.invoke(element)
+        val nestedExtractors = JavaTypeInference.getJavaBeanReadableProperties(value.getClass)
+            .map(desc => desc.getName -> desc.getReadMethod)
+            .toMap
+        new GenericInternalRow(structType.map(nestedProperty =>
+          invoke(value)(nestedExtractors(nestedProperty.name) -> nestedProperty.dataType)
+        ).toArray)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Why should we use a map here while we don't need it for the root bean?

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameSuite.java [hunk 2]
```diff
@@ -171,7 +184,12 @@ void validateDataFrameWithBeans(Bean bean, Dataset<Row> df) {
       schema.apply("d"));
     Assert.assertEquals(new StructField("e", DataTypes.createDecimalType(38,0), true,
       Metadata.empty()), schema.apply("e"));
-    Row first = df.select("a", "b", "c", "d", "e").first();
+    Assert.assertEquals(new StructField("f",
+                    DataTypes.createStructType(Collections.singletonList(new StructField(
+                            "a", IntegerType$.MODULE$, false, Metadata.empty()))),
+                    true, Metadata.empty()),
+            schema.apply("f"));
```
Comment (HyukjinKwon) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameSuite.java]: should be double spaced.

**Comment by felixcheung** [comment]
Jenkins, ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 3]
```diff
@@ -1100,13 +1101,23 @@ object SQLContext {
       attrs: Seq[AttributeReference]): Iterator[InternalRow] = {
     val extractors =
       JavaTypeInference.getJavaBeanReadableProperties(beanClass).map(_.getReadMethod)
-    val methodsToConverts = extractors.zip(attrs).map { case (e, attr) =>
-      (e, CatalystTypeConverters.createToCatalystConverter(attr.dataType))
+    val methodsToTypes = extractors.zip(attrs).map { case (e, attr) =>
+      (e, attr.dataType)
+    }
+    def invoke(element: Any)(tuple: (Method, DataType)): Any = tuple match {
```
[the remark this answers is not in the document]
Comment (michalsenkyr) [reply sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Good point. Thank you. Changed it in the latest commit

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 4]
```diff
@@ -1098,16 +1098,24 @@ object SQLContext {
       data: Iterator[_],
       beanClass: Class[_],
       attrs: Seq[AttributeReference]): Iterator[InternalRow] = {
-    val extractors =
-      JavaTypeInference.getJavaBeanReadableProperties(beanClass).map(_.getReadMethod)
-    val methodsToConverts = extractors.zip(attrs).map { case (e, attr) =>
-      (e, CatalystTypeConverters.createToCatalystConverter(attr.dataType))
+    def createStructConverter(cls: Class[_], fieldTypes: Iterator[DataType]): Any => InternalRow = {
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: nit: `Seq[DataType]` instead of `Iterator[DataType]`?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 5]
```diff
@@ -1098,16 +1098,24 @@ object SQLContext {
...
+      val methodConverters =
+        JavaTypeInference.getJavaBeanReadableProperties(cls).iterator.zip(fieldTypes)
+          .map { case (property, fieldType) =>
+            val method = property.getReadMethod
+            method -> createConverter(method.getReturnType, fieldType)
+          }.toArray
+      value => new GenericInternalRow(
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: We should check whether the `value` is `null` or not? Also could you add a test for the case?

**Review by ueshin — COMMENTED** [review]
LGTM for the current changes except for some comments.
I guess we need to support array/list of beans as you suggested, and map of beans as well.
@michalsenkyr Do you want to address them? We can do that here or in separate prs.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 4 above)
Comment (michalsenkyr) [reply sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: I used `Iterator`s instead of `Seq`s in order to avoid creating intermediate collections. However, I agree it's more concise without that.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 5 above)
Comment (michalsenkyr) [reply sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: You're right. Added. Thanks.

**Comment by michalsenkyr** [comment]
@ueshin Yes. I am already working on array/list support. Will add maps as well. It shouldn't require a rewrite now that the code is restructured, just new cases in pattern match. So I think it's ok to do in another PR.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 6]
```diff
@@ -1098,16 +1098,26 @@ object SQLContext {
...
+          .map { case (property, fieldType) =>
+            val method = property.getReadMethod
+            method -> createConverter(method.getReturnType, fieldType)
+          }
+      value =>
+        if (value == null) null
+        else new GenericInternalRow(
+          methodConverters.map { case (method, converter) =>
+            converter(method.invoke(value))
+          })
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: nit: please use braces for multi-lined if-else.

**Review by ueshin — COMMENTED** [review]
LGTM except for a nit.
@HyukjinKwon Could you take a look again please?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 6 above)
Comment (michalsenkyr) [reply sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Done

**Comment by ueshin** [comment]
I'd merge this not to block the following prs to support array/list and map of beans.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by ueshin** [comment]
Sorry, the merge script failed. Let me try again a while later.

**Comment by ueshin** [comment]
Seems like there is a merge commit in apache git https://git-wip-us.apache.org/repos/asf?p=spark.git, but not in GitHub yet.

**Comment by michalsenkyr** [comment]
Thanks! I created a new PR with array, list and map support.

**Comment by HyukjinKwon** [comment]
Hey @ueshin, sorry I was late. I'm a bit busy for these couple of weeks so please don't block by me and ignore me. Thank you for asking it to me.

**Comment by HyukjinKwon** [comment]
Looks good to me too!
